ta/aes_crypto_ta.c

```diff
@@ -271,6 +282,61 @@ static TEE_Result copy_secure_memory(uint32_t param_types, TEE_Param params[4])
   return TEE_SUCCESS;
 }
 
+#define CTR_AES_BLOCK_SIZE  16
+#define CTR_AES_IV_SIZE CTR_AES_BLOCK_SIZE
+#define CTR_AES_KEY_SIZE CTR_AES_BLOCK_SIZE
+
+struct sub_sample_t {
+    uint32_t clear_bytes;
+    uint32_t encrp_bytes;
+};
+
+static TEE_Result aes_Ctr128_Encrypt_secure(uint32_t param_types,
+                                     TEE_Param params[TEE_NUM_PARAMS])
+{
+  TEE_Result res = TEE_SUCCESS;
+  void *key, *iv, *in, *out, *iter_in, *iter_out;
+  uint32_t offset = 0;
+  struct sub_sample_t *sub_samples;
+
+  in = params[0].memref.buffer;
+  out = params[1].memref.buffer;
+  sub_samples = (struct sub_sample_t *)params[2].memref.buffer;
+  key = params[3].memref.buffer;
+  iv = (uint8_t*)key + CTR_AES_KEY_SIZE;
+  iter_in = in;
+  iter_out = out;
+
```

Comment:
Please perform some parameter checking (including checking access right on the output buffer) like it is done in aes_Ctr128_Encrypt().

Reply:
I suggest to also add a validation for :
1) buffer pointers (in, out, sub_samples and key are non-NULL)
2) buffer sizes (for in, out, sub_samples and key) to prevent accessing outside the provided memory

I may sound picky but this API is the entry point to the secure world. Input validation is critical to avoid creating a vulnerability. Without proper validation, a malicious host (NW) application could overflow the 'in' buffer to get access to some secure data or it could overflow the 'out' buffer to overwrite some secure data (keys or instructions).